### PR TITLE
Extrinsics root is calculated as part of block-building

### PR DESCRIPTION
**Pull request by gavofyork**

### CODE REVIEW

File: substrate/runtime/system/src/lib.rs [hunk 1]
```diff
@@ -68,6 +75,7 @@ decl_storage! {
 	pub BlockHash get(block_hash): b"sys:old" => required map [ T::BlockNumber => T::Hash ];
 
 	pub ExtrinsicIndex get(extrinsic_index): b"sys:xti" => required u32;
+	pub ExtrinsicData get(extrinsic_data): b"sys:xtd" => required map [ u32 => Vec<u8> ];
```
Comment (rphmeier) [inline substrate/runtime/system/src/lib.rs]: shouldn't this be cleared in `finalise`? all the values are `take`n out of it in `derive_extrinsics` but it probably leaves some residuals like length in storage.

Comment (gavofyork) [reply substrate/runtime/system/src/lib.rs]: shouldn't be possible.

Comment (gavofyork) [reply substrate/runtime/system/src/lib.rs]: it's both notionally and logically impossible. furthermore there's no way of definitively clearing a mapping.

Comment (gavofyork) [reply substrate/runtime/system/src/lib.rs]: a later PR that (re-)instates an "list" type might be reasonable though.

Comment (rphmeier) [reply substrate/runtime/system/src/lib.rs]: I don't think it's logically impossible as soon as mappings get any kind of metadata keys (for example, to enable iteration). This is something we will have to watch out for.

File: polkadot/api/src/lib.rs [hunk 2]
```diff
@@ -174,7 +177,7 @@ impl<B: Backend> PolkadotApi for Client<B, NativeExecutor<LocalDispatch>>
 	fn check_id(&self, id: BlockId) -> Result<CheckedId> {
 		// bail if the code is not the same as the natively linked.
 		if self.code_at(&id)? != LocalDispatch::native_equivalent() {
-			bail!(ErrorKind::UnknownRuntime);
+			warn!("This node is out of date. Block authoring may not work correctly.")
```
Comment (rphmeier) [inline polkadot/api/src/lib.rs]: dangerous proposition. I think authorities would prefer to get slashed for being temporarily offline than risk accidentally losing their whole bond.

also means that when we update the runtime, if >2/3 of the authorities have this behavior, they will continue growing the chain with the wrong rules.

Comment (rphmeier) [reply polkadot/api/src/lib.rs]: (I get why this change is in this PR -- the `genesis.wasm` is a nuisance to work around, but we should just stop checking this stuff in and have an environment variable for when we want to use a specific runtime wasm!)

Comment (gavofyork) [reply polkadot/api/src/lib.rs]: to be honest we're going to have to introduce semantic versioning here anyway. it's impractical to ensure that a node is updated, restarted and where necessary migrated at *exactly* the block boundary when the runtime changes, so there will need to be changeover periods. bailing here is unreasonable.

Comment (rphmeier) [reply polkadot/api/src/lib.rs]: right, but there are different approaches to that, and given that the API can change between runtimes, it's likely we might do something based on linking multiple versions of the `polkadot-api` crate. Network, consensus, and other client-level code might change between runtimes as well.

Comment (gavofyork) [reply polkadot/api/src/lib.rs]: i'd rather just have semantic versioning of the API and be done with it.

Comment (rphmeier) [reply polkadot/api/src/lib.rs]: can we agree to classify that as beyond the scope of this PR and introduce that later?

Comment (gavofyork) [reply polkadot/api/src/lib.rs]: sure

**Comment by arkpar** [comment]
What's the reason for downgrading `log`?

**Comment by gavofyork** [comment]
It was actually never upgraded, and the rest of the codebase relied on a symbol not in 0.4.

**Pull request merged**